Extra control placement 3039

Ethan-Guttman opened the pull request:
This closes the good first issue #3039 (Allow controls to be placed at top-center, bottom-center, left-center, and right-center)

example pictures showcasing all four new positions:
![image](https://github.com/mapbox/mapbox-gl-js/assets/61851595/82b4929c-6dc4-4b7c-95a2-ae829f2abe2a)
![image](https://github.com/mapbox/mapbox-gl-js/assets/61851595/3cb4feba-c28f-451f-b0d3-46111b6219f7)

Multiple ctrls will have margins between themselves like the already existing positions. However, depending on their widths in the top and bottom center they can look weird (smaller elements floating left).

I ordered the positions code in the css and map.js files by going from top-left to center-left clockwise listing each position too.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/mapbox/mapbox-gl-js?pullRequest=13104) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/mapbox/mapbox-gl-js?pullRequest=13104) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/mapbox/mapbox-gl-js?pullRequest=13104) it.</sub>

mourner commented:
Sorry for the late response — this PR is on our mind and we'll look at it closer after we're done with the current v3.3 release. 🙏

Ethan-Guttman commented:
No problem, thanks for the update!

rreusser commented on src/ui/map.js:
```diff
@@ -83,7 +83,7 @@ import type {ContextOptions} from '../gl/context.js';
 import * as TPM from '../tracked-parameters/tracked_parameters_mock.js';
 
-export type ControlPosition = 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right';
+export type ControlPosition = 'top-left' | 'top-center' | 'top-right' | 'right-center' | 'bottom-right' | 'bottom-center' | 'bottom-left' | 'left-center';
```
It seems like `left-center` and `right-center` are flipped relative to the others. Would putting the vertical positioning consistently first be easier to remember?

PR suggestion:
| | | |
|-|-|-|
| top-left | top-center | top-right |
| **left-center** | | **right-center** |
| bottom-left | bottom-center | bottom-right |

Maybe easier to remember:
| | | |
|-|-|-|
| top-left | top-center | top-right |
| **center-left** | | **center-right** |
| bottom-left | bottom-center | bottom-right |

Ethan-Guttman replied:
It's a bit of a tossup because I wanted to avoid having "center" show up as both the first and last word. Maybe make "bottom-center" into "center-bottom" too (as well as with "top-center")? Or maybe make those two "bottom-middle" and "top-middle"?

| Suggestion 1 |  | |
| ------------- | ------------- | -----------|
| top-left  | **center-top**  | top-right |
| **center-left**  |                            | **center-right** |
| bottom-left  | **center-bottom**  | bottom-right |

| Suggestion 2 |  | |
| ------------- | ------------- | -----------|
| top-left  | **top-middle** | top-right |
| **center-left**  |                            | **center-right** |
| bottom-left  | **bottom-middle**  | bottom-right |

I don't know which is best, seems like there are only tradeoffs unless we bite the bullet with my suggestion 2 and have "middle" and "center" refer to different axes

rreusser replied:
That makes sense. I know I'm engaging in some drive-by nitpicking; it just reminds me of properties I have to look up every single time, e.g. flexbox alignment vs. justification and SVG alignment-baseline center vs. central vs. middle. So I don't have a strong stance, just that some amount of logic, precedent, or memorability would be a nice goal.

rreusser replied:
I'm not sure I like this, but what about:

| Suggestion 3 |  | |
| ------------- | ------------- | -----------|
| top-left  | **top** | top-right |
| **left**  |                            | **right** |
| bottom-left  | **bottom**  | bottom-right |

Ethan-Guttman replied:
That could work and I'm leaning towards it the most. Thanks for this suggestion

rizvanovic commented:
This would be sick to get out

JabSYsEmb commented:
It seems that no changes since last comment by @Ethan-Guttman.
I can take over the issue if you're tight on time... let me know

Ethan-Guttman commented:
> It seems that no changes since last comment by @Ethan-Guttman. I can take over the issue if you're tight on time... let me know

I was waiting for a contact from the Mapbox team before updating again (since they said they would look at it closer later). I'll make the changes myself

stepankuzmin commented:
Hey @Ethan-Guttman,

Thanks for the contribution, and I'm sorry for the delay. Is there a chance you could rebase your changes on top of the latest `main`? Overall, this looks good to me, so resolving the conflicts seems to be the last thing we need to do before landing this.

Ethan-Guttman commented:
@stepankuzmin 
I'll try to rebase it by end of day today. Thanks for the go ahead!

Ethan-Guttman commented:
I'm at a bit of a loss here. This is my first ticket and I have clicked the link to sign the CLA multiple times. It might just not be reflected here as it does sometimes indicate that I have signed it when I click the link. Aside from that if there is something genuinely wrong with the mac safari version I can't test it myself (having no apple devices). Those are the only failing checks

stepankuzmin commented:
Hey @Ethan-Guttman,

The rebase seems wrong because your changes were made before we transitioned to TypeScript, and now your PR has 2k files changed. It'd be easier for maintainers to update your PR directly. Could you please make sure that you've enabled [edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) on that PR?

Ethan-Guttman commented:
@stepankuzmin 
![image](https://github.com/user-attachments/assets/702c15f6-0e62-45de-a075-8ae6ff96f16b)
Thank you for the help and let me know if anything else is needed

stepankuzmin commented:
Thanks for the contribution, @Ethan-Guttman, merging 👍
